In [5]:
import numpy as np 
import pandas as pd

In [7]:
data=pd.read_csv('/kaggle/input/mit-bih/MIT-BIH.csv')
data.head()

Unnamed: 0         0         1         2         3         4         5  \
0           0  0.006723 -0.019156  0.032603  0.006723  0.084363  0.032603   
1           1 -0.122676 -0.200316 -0.200316 -0.252076 -0.226196 -0.174436   
2           2 -0.355596 -0.407356 -0.252076 -0.277956 -0.252076 -0.277956   
3           3 -0.200316 -0.200316 -0.226196 -0.252076 -0.200316 -0.122676   
4           4 -0.174436 -0.174436 -0.174436 -0.096796 -0.096796 -0.096796   

          6         7         8  ...       171       172       173       174  \
0  0.032603  0.006723  0.006723  ... -0.536756 -0.562636 -0.614396 -0.588516   
1 -0.122676 -0.174436 -0.174436  ... -0.562636 -0.510876 -0.510876 -0.536756   
2 -0.277956 -0.329716 -0.277956  ... -0.329716 -0.355596 -0.407356 -0.433236   
3 -0.122676 -0.148556 -0.148556  ... -0.614396 -0.614396 -0.614396 -0.588516   
4 -0.200316 -0.226196 -0.174436  ... -0.459116 -0.459116 -0.510876 -0.536756   

        175       176       177       178       179  target  
0 -0.666156 -0.640276 -0.640276 -0.588516 -0.640276       1  
1 -0.562636 -0.588516 -0.588516 -0.614396 -0.562636       1  
2 -0.433236 -0.355596 -0.303836 -0.407356 -0.381476       1  
3 -0.614396 -0.743796 -0.717916 -0.743796 -0.666156       1  
4 -0.614396 -0.614396 -0.536756 -0.536756 -0.588516       1  

[5 rows x 182 columns]

In [8]:
data.shape

(107775, 182)

In [9]:
data.drop('Unnamed: 0', axis=1, inplace=True)

In [10]:
pd.options.display.float_format='{:,.3f}'.format
data.describe()

0           1           2           3           4           5  \
count 107,775.000 107,775.000 107,775.000 107,775.000 107,775.000 107,775.000   
mean       -0.144      -0.144      -0.143      -0.142      -0.141      -0.140   
std         0.538       0.538       0.537       0.536       0.536       0.536   
min       -10.305     -10.322     -10.372     -10.372     -10.372     -10.405   
25%        -0.407      -0.406      -0.406      -0.405      -0.405      -0.405   
50%        -0.171      -0.170      -0.170      -0.170      -0.169      -0.168   
75%         0.099       0.098       0.099       0.099       0.099       0.100   
max         9.516       9.370       9.370       9.487       9.545       9.633   

                6           7           8           9  ...         171  \
count 107,775.000 107,775.000 107,775.000 107,775.000  ... 107,775.000   
mean       -0.138      -0.137      -0.135      -0.133  ...       0.045   
std         0.535       0.535       0.534       0.534  ...       0.831   
min       -10.355     -10.339     -10.289     -10.256  ...      -9.400   
25%        -0.402      -0.402      -0.400      -0.399  ...      -0.486   
50%        -0.167      -0.166      -0.165      -0.160  ...       0.032   
75%         0.100       0.100       0.100       0.101  ...       0.545   
max         9.516       9.429       9.400       9.429  ...      11.971   

              172         173         174         175         176         177  \
count 107,775.000 107,775.000 107,775.000 107,775.000 107,775.000 107,775.000   
mean        0.057       0.070       0.082       0.094       0.105       0.116   
std         0.835       0.838       0.841       0.843       0.845       0.848   
min        -9.400      -9.400      -9.459      -9.400      -9.381      -9.264   
25%        -0.477      -0.466      -0.455      -0.442      -0.434      -0.424   
50%         0.048       0.060       0.073       0.080       0.091       0.103   
75%         0.560       0.577       0.593       0.607       0.620       0.637   
max        11.761      11.531      11.512      11.607      11.703      11.646   

              178         179      target  
count 107,775.000 107,775.000 107,775.000  
mean        0.127       0.138       1.300  
std         0.849       0.850       0.799  
min        -9.323      -9.303       0.000  
25%        -0.412      -0.400       1.000  
50%         0.111       0.119       1.000  
75%         0.650       0.661       1.000  
max        11.550      11.550       4.000  

[8 rows x 181 columns]

In [11]:
data.duplicated().sum()

0

In [12]:
data["target"].value_counts()

target
1    89694
2     8001
4     6487
3     2814
0      779
Name: count, dtype: int64

***The heartbeats are categorized into
89694 for Normal beats, 6487 for Ventricular ectopic beats,
2814 for Supraventricular ectopic beats, 779 for Fusion beats,
and 24 for Unknown beats.--> "Imbalanced Data"***


In [14]:
from sklearn.model_selection import train_test_split
from collections import Counter

x= data.drop('target',axis=1)
y= data['target'] 

X_train, X_test, y_train, y_test = train_test_split(
    x, y, test_size=0.25, stratify=y, random_state=42
)


print("Distribution of classes in training data:", Counter(y_train))
print("Distribution of classes in testing data:", Counter(y_test))


Distribution of classes in training data: Counter({1: 67270, 2: 6001, 4: 4865, 3: 2111, 0: 584})
Distribution of classes in testing data: Counter({1: 22424, 2: 2000, 4: 1622, 3: 703, 0: 195})


In [15]:
from imblearn.under_sampling import RandomUnderSampler
from imblearn.over_sampling import RandomOverSampler

rus = RandomUnderSampler(sampling_strategy={1: 25000}, random_state=42)
X_train_resampled, y_train_resampled = rus.fit_resample(X_train, y_train)

ros = RandomOverSampler(sampling_strategy={3: 25000, 4: 25000, 0: 25000,2:25000}, random_state=42)
X_train_balanced, y_train_balanced = ros.fit_resample(X_train_resampled, y_train_resampled)

print("Final distribution of labeled data:", Counter(y_train_balanced))


Final distribution of labeled data: Counter({0: 25000, 1: 25000, 2: 25000, 3: 25000, 4: 25000})


In [16]:
pip install imblearn

Note: you may need to restart the kernel to use updated packages.


In [17]:
from imblearn.under_sampling import RandomUnderSampler
from imblearn.over_sampling import RandomOverSampler

rus = RandomUnderSampler(sampling_strategy={1: 25000}, random_state=42)
X_train_resampled, y_train_resampled = rus.fit_resample(X_train, y_train)

ros = RandomOverSampler(sampling_strategy={3: 25000, 4: 25000, 0: 25000,2:25000}, random_state=42)
X_train_balanced, y_train_balanced = ros.fit_resample(X_train_resampled, y_train_resampled)

print("Final distribution of labeled data:", Counter(y_train_balanced))


Final distribution of labeled data: Counter({0: 25000, 1: 25000, 2: 25000, 3: 25000, 4: 25000})


In [18]:
import pandas as pd

balanced_data = pd.DataFrame(X_train_balanced)
balanced_data['target'] = y_train_balanced
print(balanced_data['target'].value_counts())

target
0    25000
1    25000
2    25000
3    25000
4    25000
Name: count, dtype: int64


/tmp/ipykernel_30/2069629483.py:4: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  balanced_data['target'] = y_train_balanced


# LSTM 

In [20]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense, Dropout
from sklearn.metrics import classification_report, confusion_matrix


In [21]:
model = Sequential([
    LSTM(128, input_shape=(125000, 125000), return_sequences=True),
    Dropout(0.2),
    LSTM(64),
    Dropout(0.2),
    Dense(64, activation='relu'),
    Dense(5, activation='softmax')  
])


/opt/conda/lib/python3.10/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


In [38]:
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

In [39]:
X_train_balanced_np = X_train_balanced.to_numpy()  
X_train_balanced_reshaped = X_train_balanced_np.reshape(X_train_balanced_np.shape[0], X_train_balanced_np.shape[1], 1)


history = model.fit(
    X_train_balanced_reshaped, 
    y_train_balanced,          
    epochs=10,
    batch_size=64,
    validation_split=0.2
)


Epoch 1/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 108s 67ms/step - accuracy: 0.5320 - loss: 1.0782 - val_accuracy: 0.0356 - val_loss: 2.4025
Epoch 2/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 104s 66ms/step - accuracy: 0.5808 - loss: 1.0012 - val_accuracy: 0.4576 - val_loss: 1.7170
Epoch 3/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 104s 66ms/step - accuracy: 0.7398 - loss: 0.6829 - val_accuracy: 0.5144 - val_loss: 1.4300
Epoch 4/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 104s 66ms/step - accuracy: 0.7807 - loss: 0.5940 - val_accuracy: 0.0000e+00 - val_loss: 2.5615
Epoch 5/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 104s 66ms/step - accuracy: 0.2635 - loss: 1.5018 - val_accuracy: 0.1347 - val_loss: 2.6081
Epoch 6/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 104s 66ms/step - accuracy: 0.6599 - loss: 0.8368 - val_accuracy: 0.6440 - val_loss: 1.1588
Epoch 7/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 104s 66ms/step - accuracy: 0.7763 - loss: 0.5648 - val_accuracy: 0.7394 - val_loss: 0.8672
Epoch 8/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 104s 66ms/step - accura

# Hybrid Model 1D CNN and GRU 

In [42]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv1D, GRU, Dropout, Dense, Flatten
from tensorflow.keras.optimizers import Adam

model = Sequential()

#  1D CNN
model.add(Conv1D(filters=64, kernel_size=5, activation='relu', input_shape=(180, 1)))  # Adjust input_shape based on data

#  GRU
model.add(GRU(units=64, return_sequences=False, activation='tanh'))

#  Dropout
model.add(Dropout(0.2))

# output layer
model.add(Dense(5, activation='softmax'))  

optimizer = Adam(learning_rate=0.001, decay=1e-6)
model.compile(optimizer=optimizer, loss='sparse_categorical_crossentropy', metrics=['accuracy'])

model.summary()


/opt/conda/lib/python3.10/site-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/opt/conda/lib/python3.10/site-packages/keras/src/optimizers/base_optimizer.py:33: UserWarning: Argument `decay` is no longer supported and will be ignored.
  warnings.warn(


Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv1d (Conv1D)                 │ (None, 176, 64)        │           384 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ gru (GRU)                       │ (None, 64)             │        24,960 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_2 (Dropout)             │ (None, 64)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 5)              │           325 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 25,669 (100.27 KB)

 Trainable params: 25,669 (100.27 KB)

 Non-trainable params: 0 (0.00 B)

In [44]:
history = model.fit(
    X_train_balanced_reshaped,  
    y_train_balanced,           
    batch_size=512,
    epochs=20,
    validation_split=0.2
)

Epoch 1/20
196/196 ━━━━━━━━━━━━━━━━━━━━ 4s 18ms/step - accuracy: 0.9799 - loss: 0.0632 - val_accuracy: 0.9434 - val_loss: 0.2097
Epoch 2/20
196/196 ━━━━━━━━━━━━━━━━━━━━ 3s 17ms/step - accuracy: 0.9806 - loss: 0.0610 - val_accuracy: 0.9198 - val_loss: 0.2862
Epoch 3/20
196/196 ━━━━━━━━━━━━━━━━━━━━ 3s 18ms/step - accuracy: 0.9755 - loss: 0.0756 - val_accuracy: 0.9300 - val_loss: 0.2431
Epoch 4/20
196/196 ━━━━━━━━━━━━━━━━━━━━ 3s 18ms/step - accuracy: 0.9778 - loss: 0.0709 - val_accuracy: 0.9410 - val_loss: 0.2109
Epoch 5/20
196/196 ━━━━━━━━━━━━━━━━━━━━ 3s 18ms/step - accuracy: 0.9821 - loss: 0.0572 - val_accuracy: 0.9366 - val_loss: 0.2345
Epoch 6/20
196/196 ━━━━━━━━━━━━━━━━━━━━ 4s 18ms/step - accuracy: 0.9817 - loss: 0.0595 - val_accuracy: 0.9402 - val_loss: 0.2120
Epoch 7/20
196/196 ━━━━━━━━━━━━━━━━━━━━ 4s 18ms/step - accuracy: 0.9830 - loss: 0.0546 - val_accuracy: 0.9309 - val_loss: 0.2714
Epoch 8/20
196/196 ━━━━━━━━━━━━━━━━━━━━ 4s 18ms/step - accuracy: 0.9790 - loss: 0.0665 - val_accu

In [47]:
import numpy as np
X_test = np.array(X_test)
X_test_reshaped = X_test.reshape((X_test.shape[0], X_test.shape[1], 1))


In [48]:
loss, accuracy = model.evaluate(X_test_reshaped, y_test, batch_size=512)
print(f"Test Loss: {loss}")
print(f"Test Accuracy: {accuracy}")


53/53 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 0.9583 - loss: 0.1429
Test Loss: 0.14174430072307587
Test Accuracy: 0.9581725001335144


# GRU 

In [49]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import GRU, Dense, Dropout
from tensorflow.keras.optimizers import Adam

model = Sequential()

model.add(GRU(64, input_shape=(X_train_balanced.shape[1], 1), return_sequences=False))

model.add(Dropout(0.2))

model.add(Dense(5, activation='softmax')) 

optimizer = Adam(learning_rate=0.001, decay=1e-6)

model.compile(optimizer=optimizer, loss='sparse_categorical_crossentropy', metrics=['accuracy'])

model.summary()


history = model.fit(
    X_train_balanced, y_train_balanced,
    epochs=40,
    batch_size=512,
    validation_split=0.2
)


/opt/conda/lib/python3.10/site-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)
/opt/conda/lib/python3.10/site-packages/keras/src/optimizers/base_optimizer.py:33: UserWarning: Argument `decay` is no longer supported and will be ignored.
  warnings.warn(


Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ gru_1 (GRU)                     │ (None, 64)             │        12,864 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_3 (Dropout)             │ (None, 64)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 5)              │           325 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 13,189 (51.52 KB)

 Trainable params: 13,189 (51.52 KB)

 Non-trainable params: 0 (0.00 B)

Epoch 1/40
196/196 ━━━━━━━━━━━━━━━━━━━━ 4s 16ms/step - accuracy: 0.3052 - loss: 1.4711 - val_accuracy: 0.3220 - val_loss: 2.0000
Epoch 2/40
196/196 ━━━━━━━━━━━━━━━━━━━━ 3s 15ms/step - accuracy: 0.6764 - loss: 0.8407 - val_accuracy: 0.4360 - val_loss: 1.8029
Epoch 3/40
196/196 ━━━━━━━━━━━━━━━━━━━━ 3s 15ms/step - accuracy: 0.7362 - loss: 0.7023 - val_accuracy: 0.4793 - val_loss: 1.6856
Epoch 4/40
196/196 ━━━━━━━━━━━━━━━━━━━━ 3s 16ms/step - accuracy: 0.7606 - loss: 0.6466 - val_accuracy: 0.4374 - val_loss: 1.8225
Epoch 5/40
196/196 ━━━━━━━━━━━━━━━━━━━━ 3s 16ms/step - accuracy: 0.7756 - loss: 0.6001 - val_accuracy: 0.5006 - val_loss: 1.5844
Epoch 6/40
196/196 ━━━━━━━━━━━━━━━━━━━━ 3s 15ms/step - accuracy: 0.7929 - loss: 0.5611 - val_accuracy: 0.4902 - val_loss: 1.5841
Epoch 7/40
196/196 ━━━━━━━━━━━━━━━━━━━━ 3s 15ms/step - accuracy: 0.8082 - loss: 0.5259 - val_accuracy: 0.5104 - val_loss: 1.4465
Epoch 8/40
196/196 ━━━━━━━━━━━━━━━━━━━━ 3s 15ms/step - accuracy: 0.8241 - loss: 0.4876 - val_accu

In [51]:
loss, accuracy = model.evaluate(X_test, y_test)
print(f"Test Loss: {loss}")
print(f"Test Accuracy: {accuracy}")

842/842 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.9418 - loss: 0.2135
Test Loss: 0.20855173468589783
Test Accuracy: 0.9434382319450378


# "The hybrid system consisting of 1D CNN and GRU achieved the highest accuracy of 95.81%."